File: morning_exercises/day13.cs
class BSTNode2(int val = 0, BSTNode2? left = null, BSTNode2? right = null)
{
    public int val = val;
    public BSTNode2? left = left;
    public BSTNode2? right = right;

    public static bool Validator(BSTNode2? root, int min, int max)
    {
        if (root==null) return true;

        if (root.val>=max || root.val<=min) return false;

        return Validator(root.left, min, root.val) && Validator(root.right, root.val, max);
    }

    public static bool IsValidBST(BSTNode2 root)
    {
        return Validator(root, int.MinValue, int.MaxValue);
    }
}

File: morning_exercises/day13.ts
class BSTNode2 {
    constructor (
        public val: number = 0,
        public left: BSTNode2 | null = null,
        public right: BSTNode2 | null = null,
    ) {}

    static validator(root: BSTNode2 | null, min:number, max:number): boolean {
        if (root==null) return true;

        if (root.val>=max || root.val<=min) return false;

        return (BSTNode2.validator(root.left, min, root.val) && BSTNode2.validator(root.right, root.val, max));
    }

    static isValidBST(root: BSTNode2 | null): boolean {
        return BSTNode2.validator(root, -Infinity, Infinity);
    }
}

{
    const root = new BSTNode2(6);
    root.left = new BSTNode2(4);
    root.right = new BSTNode2(9);
    root.left.left = new BSTNode2(2);
    root.left.right = new BSTNode2(5);

    console.log(BSTNode2.isValidBST(root));
}